### PR TITLE
build(minimal-build): allow using eBPF driver with minimal builds

### PR DESCRIPTION
**Pull request by Molter73**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area build

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:
Historically, the minimal build was only able to run using the kernel module driver, but since the addition of the modern probe this has also become an option for running with minimal builds. I believe there's value in allowing adopters of the libs to use the minimal build with the eBPF probe.

This change means that scap will always be linked against zlib and libelf, but I believe this is a worthwhile trade off.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by Molter73** [comment]
I know we are in the middle of cutting a release, just wanted to open the PR so I don't forget, we can look into this at a later time.

**Comment by Andreagit97** [comment]
Thank you @Molter73 this is really valuable! we will take a look ASAP, probably after the release :/

File: CMakeLists.txt
Comment (Andreagit97) [inline CMakeLists.txt:50]: just a minor note on the fly on which maybe we want to start discussing:
* What about having a unique `MINIMAL_BUILD`? Is not clear to me why bpf was excluded in principle :thinking: 
* What about enabling also the modern probe in `MINIMAL_BUILD`? In the end is just another ebpf driver and could be really useful for example in Falco with minimal build because we can test it even if we have no prebuilt drivers

WDYT? @falcosecurity/libs-maintainers

Comment (leogr) [reply CMakeLists.txt:50]: > * What about enabling also the modern probe in `MINIMAL_BUILD`?

:+1: 

We can also think of including `zlib` in the minimal build. I don't see it as a huge issue. wdyt?
cc @gnosek

Comment (Molter73) [reply CMakeLists.txt:50]: > * What about having a unique `MINIMAL_BUILD`? Is not clear to me why bpf was excluded in principle 🤔

AFAICT, ebpf was excluded to allow for building without linking to zlib, doesn't seem to be any other reason for this. If we can directly add eBPF into the minimal build, this PR will become a lot smaller.


> * What about enabling also the modern probe in `MINIMAL_BUILD`?

Because of how everything is setup, it's already possible to set `MINIMAL_BUILD` and `BUILD_LIBSCAP_MODERN_BPF` to `ON` and everything will be setup and work correctly, so no change is needed for this to work (at least when I tried it, it worked this way).

Comment (Andreagit97) [reply CMakeLists.txt:50]: > Because of how everything is setup, it's already possible to set MINIMAL_BUILD and BUILD_LIBSCAP_MODERN_BPF to ON and everything will be setup and work correctly, so no change is needed for this to work (at least when I tried it, it worked this way).

oh, great news! never tried! thanks

> AFAICT, ebpf was excluded to allow for building without linking to zlib, doesn't seem to be any other reason for this. If we can directly add eBPF into the minimal build, this PR will become a lot smaller.

To be honest i would go for it, so including also `zlib`, in this way we could avoid yet another cmake option :)

Comment (Molter73) [reply CMakeLists.txt:50]: Just to say this out loud, my initial intention was to split the `MINIMAL_BUILD` into multiple options that could be toggled individually or set them all with `MINIMAL_BUILD`, but desisted because it would've been a lot of work and I only needed the minimal build to run with eBPF. We can discuss this alternative approach if it's something that would be interesting for the community.

**Comment by gnosek** [comment]
TBH I'm not sure why we even have a dedicated minimal build mode in the first place. Ideally, we'd have feature flags for all the major features that MINIMAL_BUILD disables and then you'd either disable e.g. zlib (which would take eBPF away AIUI) or not, independent from other features.

Until that day comes, :+1: from me, although the checks like https://github.com/falcosecurity/libs/pull/1325/files#diff-3ef409c4a80658bffe80055929e4bfd351c6dc05877b1eb1162b19b76a3c9842R24 start to become ugly

**Comment by Molter73** [comment]
> Until that day comes, 👍 from me, although the checks like https://github.com/falcosecurity/libs/pull/1325/files#diff-3ef409c4a80658bffe80055929e4bfd351c6dc05877b1eb1162b19b76a3c9842R24 start to become ugly

Yeah, it's ugly as hell, I apologize for it.

File: CMakeLists.txt
Comment (Andreagit97) [reply CMakeLists.txt:50]: yeah until we don't have a clear use case I would avoid putting this burden on your shoulders :) If we find alternative use cases i would be happy to consider this splitting you proposed!

**Comment by dwindsor** [comment]
Hello! We definitely need something like this, as we're using the modern eBPF in our closed source application. =)

Currently, we're patching userspace/libscap/CMakeLists.txt as such, and it just works:

```
diff -purN orig/userspace/libscap/CMakeLists.txt patched/userspace/libscap/CMakeLists.txt
--- orig/userspace/libscap/CMakeLists.txt	2023-08-21 10:37:15.177875298 -0500
+++ patched/userspace/libscap/CMakeLists.txt	2023-08-21 10:54:44.366683611 -0500
@@ -21,9 +21,13 @@ option(BUILD_LIBSCAP_MODERN_BPF "Enable
 
 include(ExternalProject)
 
-if(WIN32 OR NOT MINIMAL_BUILD)
-	include(zlib)
-endif()
+#if(WIN32 OR NOT MINIMAL_BUILD OR BUILD_LIBSCAP_MODERN_BPF)
+#	include(zlib)
+#endif()
```

**Comment by Andreagit97** [comment]
If we all agree here, I think we can go with only one `MINIMAL_BUILD` option which allows:
- kmod
- legacy bpf
- modern bpf

**Comment by FedeDP** [comment]
Fully agree with Andrea! :+1: from me

**Comment by Molter73** [comment]
If the general consensus is that we should just allow all drivers to be used with the `MINIMAL_BUILD` I'll gladly change the PR, I've already started looking into this.

**Comment by Molter73** [comment]
I've changed the PR to always link scap against zlib and libelf, as well as always adding the bpf engine. IMO, it's a lot more streamlined, but let me know if you see any downsides to this approach and would like it implemented in some other way.

**Comment by FedeDP** [comment]
I love this, much simplier cmake files too.
I agree with @gnosek : sooner or later we'll want to directly expose `BUILD_LIBSCAP_ENGINE_X` options, and improve the situation. But today is not that day.
:+1: for this approach from me!

Also, we are currently frozen for the 0.13 tag, but are nearing the code thaw; if we all agree, we could merge this one right after the thaw!

**Comment by FedeDP** [comment]
> /usr/bin/ld: cannot find -l/usr/lib/x86_64-linux-gnu/libz.so: No such file or directory
> /usr/bin/ld: cannot find -l/usr/lib/x86_64-linux-gnu/libelf.so: No such file or directory

:laughing:

**Comment by Molter73** [comment]
> > /usr/bin/ld: cannot find -l/usr/lib/x86_64-linux-gnu/libz.so: No such file or directory
> > /usr/bin/ld: cannot find -l/usr/lib/x86_64-linux-gnu/libelf.so: No such file or directory
> 
> 😆

I really should've seen this coming, shouldn't I? 🤣

**Comment by FedeDP** [comment]
Hi @Molter73  can you rebase this one?

**Comment by FedeDP** [comment]
> Hi @Molter73 can you rebase this one?

It seems there is no need :) About my other comment, we can fix that while removing `UDIG` support IMHO. Not a big deal.

File: cmake/modules/libscap.cmake
Comment (FedeDP) [inline cmake/modules/libscap.cmake:91]: FYI @geraldcombs

Comment (geraldcombs) [reply cmake/modules/libscap.cmake:91]: LGTM

**Review by FedeDP — APPROVED** [review]
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Review by Andreagit97 — APPROVED** [review]
/approve

**Pull request merged**
